### ISSUE TITLE
cannot delete user "Associated data exists, please delete them first"

### ISSUE THREAD

**Issue by maddyobrienjones** [apache/superset#13345]
When trying to delete a user, the error message "Associated data exists, please delete them first" appears. All charts and dashboards created by that account have been deleted. I have also set the user as inactive and deleted its role setting. I am able to edit the user but not delete it. 

I found the following similar issue which was marked as stale with no response: https://github.com/apache/superset/issues/8752 

### Expected results

Successful deletion of a user

### Actual results

"Associated data exists, please delete them first" error message

Logs when navigating to user list and trying to delete user (domain has been replaced with deploymentlink.com):
10.4.15.23 - - [25/Feb/2021:19:48:28 +0000] "GET /users/list/ HTTP/1.1" 200 26741 "deploymentlink.com/superset/welcome" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/88.0.4324.150 Safari/537.36"
10.4.15.23 - - [25/Feb/2021:19:48:34 +0000] "POST /users/delete/3 HTTP/1.1" 302 299 "deploymentlink.com/users/list" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/88.0.4324.150 Safari/537.36"
10.4.15.23 - - [25/Feb/2021:19:48:34 +0000] "GET /users/list/ HTTP/1.1" 200 26786 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/88.0.4324.150 Safari/537.36"

#### Screenshots

![Screen Shot 2021-02-25 at 2 42 56 PM](https://user-images.githubusercontent.com/35735580/109208087-e616ce80-7777-11eb-8b28-4ff3920c773b.png)

#### How to reproduce the bug

1. Go to 'List Users'
2. Navigate to user row
3. Click on 'Delete' icon

### Environment

(please complete the following information):

- superset version: `superset version`: Superset 0.999.0dev
- python version: `python --version`: 3.7.9
- node.js version: `node -v`: couldn't find

### Checklist

Make sure to follow these steps before submitting your issue - thank you!

- [x] I have checked the superset logs for python stacktraces and included it here as text if there are any.
- [ ] I have reproduced the issue with at least the latest released version of superset.
- [x] I have checked the issue tracker for the same issue and I haven't found one similar.

### Additional context

none

**Comment by oguzhanyediel** [comment]
Any update?

**Comment by maddyobrienjones** [comment]
no updates from my end

**Comment by oguzhanyediel** [comment]
Thanks for the quick reply @maddyobrienjones

I encountered the same error a couple of days ago. I don't know what I will do because I can't remove the user although all charts and dashboards created by that user have been deleted.

By the way, the reason that I want for the user removal is the following situation.

<img width="816" alt="add_to_dashboard_error" src="https://user-images.githubusercontent.com/28841022/135750644-fe75b15a-8db9-4841-8607-ba544b248596.png">

I don't know again what I will do and I wanted to delete&add the user again. It will be greatly appreciated if anyone knows a solution to this strange situation.

**Comment by oguzhanyediel** [comment]
My second problem is solved by @nytai He told me that any user who wants to add a chart to any dashboard may have to add himself/herself as an owner of the dashboard.

The other error 'Associated data exists, please delete them first' is still ongoing.

**Comment by kushmangal** [comment]
Any update here?

**Comment by maddyobrienjones** [comment]
no updates from me. this is low priority in my backlog of tasks.

**Comment by nytai** [comment]
I think there's also a logs table. Tbh deleting a user if quite difficult since cascade hasn't been set on a lot of the foreign key relationships, so you’ll have to find them and delete the records manually. 

If possible, just mark the user as inactive and change all the fields that might cause issues with unique constraints if the same user is trying to sign up again

**Comment by nytai** [comment]
Also this stackoverflow post mentions a query you can run to find all the foreign key constraints on `ab_users.id`

**Comment by Petah** [comment]
I tried to delete a user I just made, and it doesn't work giving this error. How does a new user own anything, and how can I find and delete it?

**Comment by sdezza** [comment]
Same problem here. Really need something to be able to delete a user

**Comment by ChangbingChen** [comment]
> I think there's also a logs table. Tbh deleting a user if quite difficult since cascade hasn't been set on a lot of the foreign key relationships, so you’ll have to find them and delete the records manually.
> 
> If possible, just mark the user as inactive and change all the fields that might cause issues with unique constraints if the same user is trying to sign up again

@nytai , it's great! it works for me. When i tried to delete a user,  it took a long time and timed out(log info, [CRITICAL] WORKER TIMEOUT (pid:5783)). So i deleted the records of the logs table, and the user can be deleted normally.

**Comment by simonvanderveldt** [comment]
Has anything been done for this for 2.0.0?
Deletion should just work, after a warning everything should just automatically be removed, one shouldn't have to manually go through tabs or cleaning up records in a database (to be able to) delete a user.

**Comment by bonamim** [comment]
Hey guys,

The version 2.0.0 has the same problem ⚠️ 
We were able to reproduce the problem after a user ran a query, and then we were unable to remove this user.

It was necessary to remove the user data in this case from two tables:

- logs
- query

**Comment by ivan-price-acted** [comment]
puting this here for future use:

```
-- user to be deleted is id XXX
delete from favstar where user_id=XXX;
delete from key_value where changed_by_fk=XXX;
delete from key_value where created_by_fk=XXX;
delete from logs where user_id=XXX;
delete from ab_user where id=XXX;
```

**Comment by max-addison** [comment]
Still having this issue. The user im trying to delete has never created a chart or dashboard

**Comment by C-monC** [comment]
If you're using the latest version it may be possible that there are new tables.

Run the above script and check what the error message is for the last line. Modify the script accordingly and post it here.
I come grab this snippet whenever I need to delete a user.

**Comment by LeoDiep** [comment]
just want to save time for who don't know how to do as [C-monC](https://github.com/C-monC) above,
if you run superset by docker, the default metadata of superset is from the container name 'superset_db' and the image is prosgre. This database is already connected on Superset UI as example database when you load the examples. However you cannot delete records in this database when you haven't eddited its setting and allow DML. 

After that, just go to sqllab, find the userid you want to delete, run 'delete from ab_user where id = 5' , it will show you which table have data associated with this user_id. Then run delete query of all those related table and run 'delete from ab_user where id = 5' again, you can now remove that user.

**Comment by julizet** [comment]
I ran into the same problem. I can't delete any user even though the use was recently created and doesn't own any chart, board or dataset.

How could it be that the user deletion doesn't cascade through every data like e.g. logs?

Appreciate some solutions (two years after initial report) for how to solve this bug.

**Comment by eschutho** [comment]
Someone else reported a solution here that you might want to try: https://github.com/apache/superset/issues/2318

**Comment by beeritis** [comment]
Airflow v2.6.0

Just had the same problem - following worked for me - 

```

delete  from dag_run_note where user_id='<userid>;
delete from ab_user_role where user_id='<userid>';
delete  from ab_user where username='<username>';
```

Not a fan of force deleting but this is a fundamental bug that needs fixing.

**Comment by jaroet** [comment]
Same bug for me on 3.1.0. I have no access to the SuperSet database so I will set the user as inactive. But it should really be fixed.

**Comment by ziggekatten** [comment]
This is driving me nuts as well. I do not feel comfortable doing deletes outside of the ORM, as it might wreck stuff and make upgrades fail

**Comment by xavier-GitHub76** [comment]
Hello,
I have already create and delete users without problem.
We ca consider the cascade delete is existing but for an unknown reason, I have also this issue on superset v3.
![image](https://github.com/apache/superset/assets/82046143/af5c9de4-cd73-4eb3-bbe2-32bd7fc6957a)

The message should detail the data to remove.
I will test :

> puting this here for future use:
> 
> ```
> -- user to be deleted is id XXX
> delete from favstar where user_id=XXX;
> delete from key_value where changed_by_fk=XXX;
> delete from key_value where created_by_fk=XXX;
> delete from logs where user_id=XXX;
> delete from ab_user where id=XXX;
> ```

**Comment by xavier-GitHub76** [comment]
Hello,
I systematically reproduce the problem "impossible to delete a user : Associated data exists, please delete them first".

I have identified the cause: This only concerns users who have created elements in Superset. Despite the deletion of elements created by a user from the graphical user interface, the deletion of the user is impossible because the "key_value" table is not cleaned at the time of deletion, this creates an integrity constraint preventing the deletion of the user.

The other tables (logs, slices, slice_user) are correctly updated.

The “key_value” table has a relationship only with the “ab_user” table.
![image](https://github.com/apache/superset/assets/82046143/4530d270-0101-4039-b436-73ea142ee970)

It is possible to work around the problem by executing a delete query on the key_value table :
delete from public.key_value where created_by_fk = <id_user_to_delete>

Testing the workaround : 
"Deleted Row" 👍 

The following tables are well updated:
    • ab_user.sql
        ◦ Deletion of the line relating to the user
    • ab_user_role.sql
        ◦ Deletion of the line relating to the user
    • logs.sql
        ◦ Lines mentioning <user id>  now mention user = None (like a user deleted without objects created)

Best regards

**Comment by rusackas** [comment]
Interesting, thanks for the context. @michael-s-molina or @villebro would know best about the intricacies of deleting `key_value` and whether it's preferable to remove them along with the user, or leave them orphaned (i.e. if they're needed by other users in other contexts, like sharing a URL with someone)

**Comment by villebro** [comment]
@rusackas @michael-s-molina @xavier-GitHub76 IMO orphaning the rows is definitely the right course of action here, as permalinks etc should continue existing after a user record is deleted. And for other temporary key value pairs, there's usually an expiration date attached to them, which will ensure that the records disappear in due course.

**Comment by xavier-GitHub76** [comment]
So the best practice would be to wait until the expiration date has passed before being able to delete the user ?
If so, several questions arise: 
- How can an administrator find out about this date ?
- What is the mechanism that deletes keys that have become obsolete ?
- Is it possible to speed up this deletion ?

Thank you for any details you can provide us.

**Comment by kainchow** [comment]
> puting this here for future use:
> 
> ```
> -- user to be deleted is id XXX
> delete from favstar where user_id=XXX;
> delete from key_value where changed_by_fk=XXX;
> delete from key_value where created_by_fk=XXX;
> delete from logs where user_id=XXX;
> delete from ab_user where id=XXX;
> ```

I would like to know what the key_value table actually does

**Comment by xavier-GitHub76** [comment]
The alert message "Associated data exists, please delete them first" could be replaced by a more appropriate message.
It is requested to delete the elements preventing deletion but the user cannot identify them or even cannot act (eg: key_value)

The message could be : 
"Unable to delete this user
Traces concerning him are present (e.g. creation of elements on his part)
Please use account deactivation"

**Comment by rusackas** [comment]
This has been silent for > 6 months. I'm wondering if anyone here has interest in opening a PR (updating the error message or otherwise), but if not, it may eventually be closed as stale/not-planned.

**Comment by ivan-price-acted** [comment]
speaking for myself this was only something we were doing in an exotic use case (migration from one auth provider to another), so the sql approach was ok for us. normally I would agree that disabling users is preferable to deleting them.

**Comment by knightmre** [comment]
On superset 4.1.1, remove all roles from the user then use this SQL Query: 

```sql
-- user to be deleted is id XXX
delete from query where user_id=XXX;
delete from logs where user_id=XXX;
delete from tab_state where user_id=XXX;
delete from favstar where user_id=XXX;
delete from key_value where changed_by_fk=XXX;
delete from key_value where created_by_fk=XXX;
delete from logs where user_id=XXX;
delete from ab_user where id=XXX;
```

**Comment by gelila-s** [comment]
Hi, my name is Gelila and I have a group that is interested in working on this issue. I've reproduced the issue and confirmed that a user that has created a dashboard cannot be deleted, even after deleting that user's dashboard, with the error message "Associated data exists, please delete them first." We'd love to try fixing it if a fix is still needed.

**Comment by rusackas** [comment]
@gelila-s that'd be great! I'll just ping @dpgaspar @michael-s-molina @villebro @mistercrunch here in case they have feelings about the correct approach here. 

I'm not personally close to the issue at hand (rarely do I delete a user, if ever) but I think some general things to keep in mind would be about being extra cautious/safe here. 
* Might it warrant an extra confirmation (i.e. "You'll be removing {stuff here}. Type DELETE to continue") since this may be unsafe
* If any of the records to be deleted have value to other users (which doesn't seem to be the case) it would be good to reassign them to another user. Sounds like this isn't a concern here though.
* I'm _not_ sure if some things like `logs` have use for certain orgs, like a permanent audit trail for legal/governance purposes. Maybe they're safe to delete, but if not, it _seems_ possible to update the log's `user_id` from `X` to `DELETED_X`?

It's a whole other can of worms (and a much larger project) but eventually I would love to see more (or all!) tables in Superset support a "soft delete" where a new column in the table (i.e. a boolean "deleted" column) gets set, and those deleted records are effectively filtered out and ignored throughout Superset. Then they'd be recoverable in an emergency... like if you bring that user back later, or someone accidentally deletes a dashboard, your DBA could go in and hand-tweak the record(s) to restore things.

Anyway, that's a long-winded way of saying "heck yeah, go for it!" but just erring on the side of caution when/where it makes sense.

**Comment by mistercrunch** [comment]
Would be great to have some reassignment logic, at least as an option. Often when someone leaves an organization, the policy is to re-assign their assets to their manager or a team member.  Feels like a better default to me.

**Comment by jl141** [comment]
Hi! @rusackas @mistercrunch @dpgaspar @michael-s-molina @villebro 
We have a solution proposal as follows and would greatly appreciate feedback about our approach:
- Add a new boolean column `DELETED` in table `ab_user`
- Only admins can delete users; during this process (if there are dependencies preventing the deletion) they will be given the option to either (1) reassign the user’s associated data to another user (or themselves), or (2) orphan the associated data so that permalinks still work
- Deletion confirmation e.g. "You will be deleting user {username}. Type DELETE to continue" 
- Modify the appropriate database entries:
  - For (1), we would modify relevant entries with the deleted user’s ID to be the reassigned user’s ID (i.e. dashboards, slices)
  - For both, we would modify the column `DELETED` of the table `ab_user` to be `TRUE`. 
- The deleted user will no longer be able to log in or perform any actions. We can achieve this by setting is_active to `FALSE`
- The deleted user’s username “X” will instead be updated to “[DELETED]_X”, for the purpose of logs, permalinked dashboards, or account recreation under the same username. A similar modification will be done to the user’s email field. 
- Future queries in the Superset app will filter out rows where the user ID corresponds to a deleted user. Thus a deleted user’s dashboards and charts will only be accessible through permalinks, unless they have been reassigned.

**Comment by michael-s-molina** [comment]
@jl141 Soft deletion of users with a DELETED column was previously discussed in [SIP-27](https://github.com/apache/superset/discussions/18386) and might require substantial changes to logic that fetches data. Superset is evolving to a more team-based ownership model (check [SIP-152](https://github.com/apache/superset/issues/32116)) which leads me to think that making reassignment required could be a simpler solution and more close to what we have seen in practice. At Airbnb, we generally receive the request to reassign assets such as dashboards, charts and saved queries to other team members once someone leaves the company and forget to set new owners.

**Comment by jl141** [comment]
@michael-s-molina Thanks for the reply! We've looked into the items you brought up and have modified our approach:
- When an admin deletes a user, we will require them to choose a user to reassign assets (dashboards, charts, saved queries) to, given that there exists associated data. 
- Regarding other data, we will either set the delete rule on `ab_user.id` to be `CASCADE` or `SET NULL`, depending on whether they are saved in the tables as `user_id` or `changed/created/last_saved_by_fk`. e.g. `CASCADE` on tables `ab_user_role`, `favstar`, `query`, `sql_metrics`, etc. and `SET NULL` on tables `ab_user`, `annotation`, `annotation_layer`, `dbs`, etc.
- Regarding log data integrity in the context of deleting a user: The logs table contains references to user IDs that seemingly cannot be broken. Thus in the case of the hard deleting a user, we are unsure how to achieve this selective deletion. 
  - One option is to `CASCADE` delete the logs along with the user, but this is not ideal if the logs are needed for later reference or legal purposes. 
  - Another option is to soft delete the user instead by keeping the deleted user in the `ab_user` table, and changing its username from “X” to “[DELETED]_X”, which would keep the logs intact while allowing the username “X” to be reused. However, this would also mean that the other data (as in the bullet above) would neither be deleted nor set null, and we may run into the aforementioned problem of substantial logic changes. 
  - An additional option (utilizing `SET NULL` delete rule on `ab_user.id`) is to add a column `user_info` to the logs table that provides information about the user (e.g. username + email) and logic to access this information when it turns out that `user_id` is null.

**Comment by michael-s-molina** [comment]
@jl141 Yes, good point about the logs table. We definitely can't lose traceability here. We'll probably need a middle ground solution where we reassign assets ownership but preserve referenced users in logs somehow. Any thoughts here @rusackas @villebro @mistercrunch? This would be a great topic for a Town Hall or SIP review meeting.

**Comment by mistercrunch** [comment]
Right. Deleted flag + optional/configurable reassignment sound right. At Preset we rewire logs elsewhere, but could see many options as to how to treat deletion being valid for different use cases. For instance GDPR probably requires deleting or obfuscating PII and log events. 

Seems we need an endpoint that can provide all this, and maybe policies/configuration around it.

Trying to list out options:
- soft delete, keep user info
- soft delete, obfuscate PII
- true delete, also delete logs

This in combination with reassignment options:
- reassign object ownership to a single user
- delete owned objects (seems dangerous)
- orphan owned object, might create issue where ownership is required (?)

I'm probably overthinking all this. Might be able to support 95%+ of things a few option combinations

**Comment by mistercrunch** [comment]
Side note on the logs/audit table. Generally probably good practice to not use the metadata DB for this in larger environments, but safe to assume most deployments out there do use it. Reason is it puts load on the DB, and compliance-type concerns like we're discussing here. Also leaves you with an ever-growing, append only table you should trim/archive periodically.
